### PR TITLE
feature: Redis server TLS support

### PR DESCRIPTION
**Pull request by sicet7**
# Reason for This PR

We are (at my employer [ASG-Digital](https://github.com/ASG-Digital)) using Azure's product `Azure Cache for Redis` and we were having some issues connecting to their default configuration which disables non-tls connections and uses their own CA to sign their certificates. 
I implemented the required changes to be able to connect to Azure Cache for Redis default configuration without having to re-enable non-tls connections in the Azure control panel.

## Description of Changes

I added a `tls` configuration section under the redis configuration to allow for configuration of the `min_version` and `ca_file` options.
The `min_version` option configures the minimum TLS version that can be used for connecting to the remote redis server. It  accepts the following values: `1.0`, `1.1`, `1.2` and `1.3`, it defaults to `1.3` if only the `ca_file` option is provided.
The `ca_file` is a path either global or relative, to a file containing one or more PEM certificates, that should be used to verify the server certificate.

These changes are made so that existing configurations should still work without any changes needed.
The TLS configuration is only applied to the connection configuration if one or more of the configuration options `min_version` or `ca_file`  is provided.

Snipped from the RoadRunner configuration i used for testing
```yml
#Cache
kv:
    redis:
        driver: redis
        config:
            addrs:
                - "${REDIS_HOST:-127.0.0.1}:${REDIS_PORT:-6379}"
            username: "${REDIS_USERNAME}"
            password: "${REDIS_PASSWORD}"
            db: "${REDIS_DB:-0}"
            tls:
                ca_file: ../.docker/certs/Azure.crt.pem
```

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the MIT license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
       _I was not able to find any documentation in need of updating in this repo._
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
       _I was not able to find the mentioned file in this repo._
- [x] All added/changed functionality is tested.

## Summary by CodeRabbit

- **New Features**
	- Introduced TLS configuration options for enhanced security in Redis connections.
	- Added support for specifying a CA file for certificate handling.
	- New `TLSConfig` struct created to manage TLS settings, including certificate paths.
	- Enhanced `NewRedisDriver` function to utilize TLS settings for secure connections.

- **Bug Fixes**
	- Improved error handling for TLS configuration, ensuring better feedback on certificate issues.

### CODE REVIEW

**Comment by rustatian** [comment]
Hey @sicet7 👋 
Thank you for the PR 👍👍 This is indeed a cool functionality you've added 👍 

Just a few notes:
1. TLS min version should be 1.2, and it is not needed to expose this configuration option, because even for the Microsoft, TLS 1.0 and 1.1 are deprecated: [link](https://learn.microsoft.com/en-us/lifecycle/announcements/tls-support-ending-10-31-2024).
2. It'd be nice to have unified options across the RR configuration, for example, these are TLS options: [link](https://github.com/roadrunner-server/http/blob/master/servers/https/config.go#L45), you may check validation mechanism and `tls.Config` creation: [link](https://github.com/roadrunner-server/http/blob/master/tlsconf/default_tlsconf.go), and you may generally have some inspiration from this code: [link](https://github.com/roadrunner-server/http/tree/master/servers/https)
3. It would be nice if you'd be able to add tests, they're located in the `tests` folder. For TLS, I usually use `mkcert` in the GH Actions: [link](https://github.com/roadrunner-server/http/blob/master/.github/workflows/linux.yml#L72)

File: kv/config.go [hunk 1]
```diff
@@ -26,6 +26,11 @@ type Config struct {
 	IdleTimeout      time.Duration `mapstructure:"idle_timeout"`
 	IdleCheckFreq    time.Duration `mapstructure:"idle_check_freq"`
 	ReadOnly         bool          `mapstructure:"read_only"`
+	TLSConfig        TLSConfig     `mapstructure:"tls"`
```
Comment (rustatian) [inline kv/config.go:29]: Should be a pointer: `*TLSConfig` to distinguish between empty and non set values.

File: kv/config.go [hunk 2]
```diff
@@ -26,6 +26,11 @@ type Config struct {
...
+}
+
+type TLSConfig struct {
+	CaFile string `mapstructure:"ca_file"`
```
Comment (rustatian) [inline kv/config.go:33]: Also, please, follow the standard RR configuration naming for the TLS options, example: https://github.com/roadrunner-server/grpc/blob/master/tests/configs/.rr-grpc-rq-tls-rootca.yaml#L20

File: kv/driver.go [hunk 3]
```diff
@@ -74,7 +77,37 @@ func NewRedisDriver(log *zap.Logger, key string, cfgPlugin Configurer, tracer *s
 		RouteByLatency:   d.cfg.RouteByLatency,
 		RouteRandomly:    d.cfg.RouteRandomly,
 		MasterName:       d.cfg.MasterName,
-	})
+	}
+
+	if d.cfg.TLSConfig.CaFile != "" {
```
Comment (rustatian) [inline kv/driver.go:82]: Check for nil

File: kv/driver.go [hunk 4]
```diff
@@ -74,7 +77,37 @@ func NewRedisDriver(log *zap.Logger, key string, cfgPlugin Configurer, tracer *s
...
+			return nil, errors.E(op, crtReadErr)
+		}
+
+		if !rootCAs.AppendCertsFromPEM(bytes) {
+			return nil, errors.E(op, errors.Errorf("failed to parse certificates from PEM file '%s'. Please ensure the file contains valid PEM-encoded certificates", d.cfg.TLSConfig.CaFile))
+		}
+
+		tlsConfig.RootCAs = rootCAs
+		redisOptions.TLSConfig = tlsConfig
+	}
```
Comment (rustatian) [inline kv/driver.go:108]: It is better to move TLS configuration to the separate package

Comment (sicet7) [reply kv/driver.go:108]: sorry i am new to go, when you say new package do you mean in another repo or simply in another directory?? :sweat_smile:

Comment (rustatian) [reply kv/driver.go:108]: Sure, np, I'll help you 😃 
Just a new file called `tlsconfig.go` or similar :)

**Comment by sicet7** [comment]
@rustatian Some thing like this?
https://github.com/ASG-Digital/rr-go-plugin-redis/pull/1
Above is just a temp PR, that i wont merge but cherry-pick to keep proper commit history.
I simply made it to show you before i push :-)

**Comment by rustatian** [comment]
Yeah 👍 
If you don't mind, I'd jump into your PR and correct some things 😃

**Comment by sicet7** [comment]
> Yeah 👍 If you don't mind, I'd jump into your PR and correct some things 😃

I don't mind at all, go right ahead :-)

**Comment by rustatian** [comment]
A few updates: https://github.com/roadrunner-server/redis/commit/d90dce6b079b81a661f74865a4b8efcbcba003db

**Comment by rustatian** [comment]
Have you tried testing this PR on your `Azure Cache for Redis` instance?

**Comment by sicet7** [comment]
ahh...  
I pushed to the wrong branch :sweat_smile:
But no, 
currently it is not working, but i just started debugging i get the following error:


```
app-1            | panic: runtime error: invalid memory address or nil pointer dereference [recovered]
app-1            |      panic: runtime error: invalid memory address or nil pointer dereference [recovered]
app-1            |      panic: runtime error: invalid memory address or nil pointer dereference
app-1            | [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x68254d]
app-1            |
app-1            | goroutine 136 [running]:
app-1            | go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.deferwrap1()
app-1            |      go.opentelemetry.io/otel/sdk@v1.32.0/trace/span.go:422 +0x25
app-1            | go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0007fe000, {0x0, 0x0, 0xc00079ac40?})
app-1            |      go.opentelemetry.io/otel/sdk@v1.32.0/trace/span.go:455 +0xaf4
app-1            | panic({0x19dcf20?, 0x2f536d0?})
app-1            |      runtime/panic.go:785 +0x132
app-1            | go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.deferwrap1()
app-1            |      go.opentelemetry.io/otel/sdk@v1.32.0/trace/span.go:422 +0x25
app-1            | go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0007fe1e0, {0x0, 0x0, 0x0?})
app-1            |      go.opentelemetry.io/otel/sdk@v1.32.0/trace/span.go:455 +0xaf4
app-1            | panic({0x19dcf20?, 0x2f536d0?})
app-1            |      runtime/panic.go:785 +0x132
app-1            | crypto/tls.(*clientHandshakeStateTLS13).sendClientCertificate(0xc0007f8578)
app-1            |      crypto/tls/handshake_client_tls13.go:787 +0x16d
app-1            | crypto/tls.(*clientHandshakeStateTLS13).handshake(0xc0007f8578)
app-1            |      crypto/tls/handshake_client_tls13.go:148 +0x785
app-1            | crypto/tls.(*Conn).clientHandshake(0xc000532388, {0x205bfd0, 0xc000f8a4b0})
app-1            |      crypto/tls/handshake_client.go:366 +0x845
app-1            | crypto/tls.(*Conn).handshakeContext(0xc000532388, {0x205c040, 0xc0009a4620})
app-1            |      crypto/tls/conn.go:1568 +0x3a6
app-1            | crypto/tls.(*Conn).HandshakeContext(...)
app-1            |      crypto/tls/conn.go:1508
app-1            | crypto/tls.dial({0x205bf28?, 0x2fc8020?}, 0xc000a389e8, {0x1c96b58, 0x3}, {0xc000046ff0, 0x2d}, 0xc000924540)
app-1            |      crypto/tls/tls.go:159 +0x3a5
app-1            | crypto/tls.DialWithDialer(0xc000a38a18?, {0x1c96b58?, 0x203ca98?}, {0xc000046ff0?, 0xc00091a060?}, 0x0?)
app-1            |      crypto/tls/tls.go:119 +0x3d
app-1            | github.com/redis/go-redis/v9.(*Options).init.NewDialer.func1({0x205bf98?, 0xc000923c80?}, {0x1c96b58?, 0x0?}, {0xc000046ff0?, 0x0?})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/options.go:243 +0x9e
app-1            | github.com/redis/go-redis/v9.(*baseClient).dial(...)
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:397
app-1            | github.com/redis/go-redis/extra/redisotel/v9.(*tracingHook).DialHook.func1({0x205bf98?, 0xc000923c50}, {0x1c96b58, 0x3}, {0xc000046ff0, 0x2d})
app-1            |      github.com/redis/go-redis/extra/redisotel/v9@v9.7.0/tracing.go:97 +0x124
app-1            | github.com/redis/go-redis/extra/redisotel/v9.(*metricsHook).DialHook.func1({0x205bf98, 0xc000923c50}, {0x1c96b58, 0x3}, {0xc000046ff0, 0x2d})
app-1            |      github.com/redis/go-redis/extra/redisotel/v9@v9.7.0/metrics.go:194 +0xc3
app-1            | github.com/redis/go-redis/v9.(*hooksMixin).dialHook(0x0?, {0x205bf98?, 0xc000923c50?}, {0x1c96b58?, 0x45b01f?}, {0xc000046ff0?, 0xc000a38e18?})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:181 +0x119
app-1            | github.com/redis/go-redis/v9.newConnPool.func1({0x205bf98?, 0xc000923c50?})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/options.go:516 +0x35
app-1            | github.com/redis/go-redis/v9/internal/pool.(*ConnPool).dialConn(0xc000a00510, {0x205bf98?, 0xc000923c50?}, 0x1)
app-1            |      github.com/redis/go-redis/v9@v9.7.0/internal/pool/pool.go:213 +0xbc
app-1            | github.com/redis/go-redis/v9/internal/pool.(*ConnPool).newConn(0xc000a00510, {0x205bf98, 0xc000923c50}, 0x1)
app-1            |      github.com/redis/go-redis/v9@v9.7.0/internal/pool/pool.go:178 +0x178
app-1            | github.com/redis/go-redis/v9/internal/pool.(*ConnPool).Get(0xc000a00510, {0x205bf98, 0xc000923c50})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/internal/pool/pool.go:293 +0x167
app-1            | github.com/redis/go-redis/v9.(*baseClient)._getConn(0xc00100e580, {0x205bf98, 0xc000923c50})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:260 +0x32
app-1            | github.com/redis/go-redis/v9.(*baseClient).getConn(0xc00100e580, {0x205bf98?, 0xc000923c50?})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:248 +0x65
app-1            | github.com/redis/go-redis/v9.(*baseClient).withConn(0xc00100e580, {0x205bf98, 0xc000923c50}, 0xc0007f9088)
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:381 +0x49
app-1            | github.com/redis/go-redis/v9.(*baseClient)._process(0xc00100e580, {0x205bf98, 0xc000923c50}, {0x206d3c0, 0xc0009a45b0}, 0xc000538738?)
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:436 +0xd0
app-1            | github.com/redis/go-redis/v9.(*baseClient).process(0xc00100e580, {0x205bf98, 0xc000923c50}, {0x206d3c0, 0xc0009a45b0})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:405 +0x76
app-1            | github.com/redis/go-redis/extra/redisotel/v9.(*tracingHook).ProcessHook.func1({0x205bf98, 0xc000923bf0}, {0x206d3c0, 0xc0009a45b0})
app-1            |      github.com/redis/go-redis/extra/redisotel/v9@v9.7.0/tracing.go:128 +0x81e
app-1            | github.com/redis/go-redis/extra/redisotel/v9.(*metricsHook).ProcessHook.func1({0x205bf98, 0xc000923bf0}, {0x206d3c0, 0xc0009a45b0})
app-1            |      github.com/redis/go-redis/extra/redisotel/v9@v9.7.0/metrics.go:211 +0x9f
app-1            | github.com/redis/go-redis/v9.(*hooksMixin).processHook(...)
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:185
app-1            | github.com/redis/go-redis/v9.(*Client).Process(0xc000a39528?, {0x205bf98?, 0xc000923bf0?}, {0x206d3c0, 0xc0009a45b0})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/redis.go:697 +0x38
app-1            | github.com/redis/go-redis/v9.cmdable.Get(0xc0010156e0, {0x205bf98, 0xc000923bf0}, {0xc000f8e620, 0x4})
app-1            |      github.com/redis/go-redis/v9@v9.7.0/string_commands.go:54 +0x10c
app-1            | github.com/roadrunner-server/redis/v5/kv.(*Driver).MGet(0xc0009236e0, {0xc0009a2b80, 0x1, 0x5?})
app-1            |      github.com/roadrunner-server/redis/v5@v5.0.0-20241120095503-6f6e25f9ed03/kv/driver.go:169 +0x403
app-1            | github.com/roadrunner-server/kv/v5.(*rpc).MGet(0xc0008f11e0, 0xc000518af0, 0xc000d3a9c0)
app-1            |      github.com/roadrunner-server/kv/v5@v5.0.0-20240912131252-a75c68b5d775/rpc.go:109 +0x5f0
app-1            | reflect.Value.call({0xc000926480?, 0xc0004fbf50?, 0x13?}, {0x1c9745d, 0x4}, {0xc000957ef8, 0x3, 0x3?})
app-1            |      reflect/value.go:581 +0xca6
app-1            | reflect.Value.Call({0xc000926480?, 0xc0004fbf50?, 0xc00097a6a0?}, {0xc00097a6f8?, 0x10?, 0x1aceea0?})
app-1            |      reflect/value.go:365 +0xb9
app-1            | net/rpc.(*service).call(0xc0008fe580, 0xc0008fe380, 0xc000f8e508, 0xc000f8e560, 0xc000928700, 0xc00100e7c0, {0x1b27a00?, 0xc000518af0?, 0x205c4e8?}, {0x1addba0, ...}, ...)
app-1            |      net/rpc/server.go:381 +0x209
app-1            | created by net/rpc.(*Server).ServeCodec in goroutine 135
app-1            |      net/rpc/server.go:478 +0x3d7
```

**Comment by rustatian** [comment]
Ah, you may try to temporarily comment `GetClientCertificate` TLSConfig option. It should be nil if not provided any cert and key.

**Comment by sicet7** [comment]
I made it work with these changes:
https://github.com/ASG-Digital/rr-go-plugin-redis/commit/cde1077a4e8cc6041824b4b390a61cd69f6cd03a
But i have no idea if that is the correct way of doing it :sweat_smile:

**Comment by rustatian** [comment]
Yeah, this is a correct change. Because `TLSConfig` should receive nil function (the same as commenting out that option or just not providing it) instead of receiving the nil certificate on non-nil function 👍

**Comment by sicet7** [comment]
I have cherry pick it in now, i tried to force push with a new cherry pick for your commit to see if i could get it to not complain about the commit being "Unverified" :sweat_smile:

**Comment by rustatian** [comment]
Not a problem 😃 
It would be nice if you'd be able to make a PR into docs `roadrunner-server/docs` repository with the updated configuration and new functionality.
And at the same time, I'll try to push another change, to test this new functionality in the CI.

**Comment by sicet7** [comment]
Will do, is there any perticular branch that i should make the merge towards on the `roadrunner-server/docs` repo?
Or is master fine?

**Comment by rustatian** [comment]
Sorry for the late response, yeah, master is ok, thank you 👍

**Comment by rustatian** [comment]
Thank you @sicet7 👍 
I'll create anouther PR with tests update.

**Pull request merged**